internal/tools/update-go-azure-sdk/main.go

```diff
@@ -0,0 +1,729 @@
+	}
+	worktree, err := repo.Worktree()
+	if err != nil {
+		return fmt.Errorf("obtaining worktree: %+v", err)
+	}
+
+	logger.Trace(fmt.Sprintf("Performing `git reset --hard` in %s..", workingDirectory))
+	opts := &git.ResetOptions{
+		Mode: git.HardReset,
+	}
+	if err := worktree.Reset(opts); err != nil {
+		return fmt.Errorf("performing reset: %+v", err)
+	}
+
+	// Whilst the git library exposes a `clean`, it doesn't support `clean -xdf`
+	// so we'll have to shell out for that
+	logger.Trace(fmt.Sprintf("Performing `git clean` in %s..", workingDirectory))
+	args := []string{
+		"clean",
+		"-xdf",
+	}
+	cmd := exec.CommandContext(ctx, "git", args...)
+	cmd.Dir = workingDirectory
+	_ = cmd.Start()
+	_ = cmd.Wait()
+	return nil
+}
+
+func shouldUpdateFrom(existing string, new string) bool {
+	if strings.EqualFold(existing, new) {
+		logger.Trace("Versions are the same")
+		return false
+	}
+	lowered := strings.ToLower(new)
+	if strings.Contains(lowered, "alpha") {
+		logger.Trace("Is an alpha..")
+		return false
+	}
+	if strings.Contains(lowered, "beta") {
+		logger.Trace("Is a beta..")
+		return false
+	}
+	if strings.Contains(lowered, "preview") {
+		logger.Trace("Is a preview..")
+		return false
+	}
+
+	return new > existing
+}
+
+func runMakeTest(ctx context.Context, workingDirectory string) error {
+	absPath, err := filepath.Abs(workingDirectory)
+	if err != nil {
+		return fmt.Errorf("determining absolute path for %q: %+v", workingDirectory, err)
+	}
+
+	logger.Trace("Running `make test`..")
+	args := []string{
+		"test",
+	}
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	cmd := exec.CommandContext(ctx, "make", args...)
+	cmd.Dir = absPath
+	cmd.Stderr = &stderr
+	cmd.Stdout = &stdout
+	_ = cmd.Start()
+	_ = cmd.Wait()
+	if stderr.Len() > 0 {
+		output := fmt.Sprintf(`stdout:
+---
+%s
+---
+
+stderr:
+---
+%s
+---
+`, stdout.String(), stderr.String())
+		return fmt.Errorf("running `make test` in %q: %s", workingDirectory, output)
+	}
+
+	return nil
+}
+
+type updatedServiceSummary struct {
+	serviceName     string
+	olderApiVersion string
+	newApiVersion   string
+	error           *string
+}
+
+func (s updatedServiceSummary) successful() bool {
+	return s.error == nil
+}
+
+func determineApiVersionsCurrentlyUsedForService(ctx context.Context, workingDirectory string, serviceName string) (*[]string, error) {
+	absPath, err := filepath.Abs(workingDirectory)
+	if err != nil {
+		return nil, fmt.Errorf("determining absolute path for %q: %+v", workingDirectory, err)
+	}
+
+	logger.Debug(fmt.Sprintf("Determining the API Versions used for `hashicorp/go-azure-sdk` Service %q..", serviceName))
+	diffArgs := []string{
+		"-hR",
+		fmt.Sprintf("github.com/hashicorp/go-azure-sdk/resource-manager/%s/", serviceName),
+		"--include=*.go",
+		"./internal/services/..",
+	}
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	cmd := exec.CommandContext(ctx, "grep", diffArgs...)
+	cmd.Dir = absPath
+	cmd.Stderr = &stderr
+	cmd.Stdout = &stdout
```

Comment:
Would it be more reliable / less forking to use `astutil.Imports()` for this instead?

Reply:
Not a bad shout - will take a look.